### DISCUSSION TITLE
Is it possible to customize the error message for the missing option value?

### DISCUSSION THREAD

**Discussion by vadz** [CLIUtils/CLI11#814] [Q&A] [answered]
Hello,

 I've just started using CLI11 and I have a pretty simple command line utility which does
```cpp
    app.add_option("--dsn", opts.dsn, "The database connection string to use")
      ->option_text("<DSN>");
```
and works fine, except that if I run it with just `--dsn`, but without any argument, I get the following error message
```console
--dsn: 1 required TEXT missing
Run with --help for more information.
```
and the first line here seems rather awkward to me, I'd rather output something like "Required value missing for the --dsn option" or maybe "Expected --dsn option value not found".

Is it possible to customize the error message in the option definition or in some other way? Looking at `App::_parse_arg()` code I don't see any possibility to customize it, but perhaps I'm missing something?

Thanks in advance!

**Comment by phlptp** [comment]
The `TEXT` portion of that description can be modified with the use the `type_name(<string>)` or `type_name_fn(std::function<std::string()> typefun)` this can allow some customization of the help message.  

If you wanted to fully customize it, the best way would probably be to write a callback yourself that does the check,  then you can fully control the error message and all the checks.
``` c++
app.add_option("--dsn", [&opts](const results_t &results){
    if (<some check>){
           opts.dsn=results[0];
           return true}
     else{
           throw an error with some description;
            return false;
    }}, "The database connection string to use")
      ->option_text("<DSN>")->expected(0,1);
```

**Reply by vadz** [reply]
Thanks a lot, I didn't realize I could change the type string like this. Now I get `--dsn: 1 required database connection string missing` which is still a bit strange, but clear enough, I think.

But trying the second approach, with the callback, I can't make it work: the callback is not being called at all because the same exception as before is thrown from `_parse_arg()` -- except now it results in an even worse error message because there is no type name, so it's just `--dsn: 1 required  missing`.

Is the callback and/or validator passed to `check()` supposed to be called if there is nothing else following the option on the command line? I'd like this to be the case (with the empty string in the latter case), but it doesn't seem to actually happen.

**Comment by phlptp** [comment, marked as answer]
The expected(0,1) is critical there.  Without that the missing argument error is triggered.   

The `check(...)` won't really help since that is for validating individual arguments.  Though I suppose that might work as well if you checked for a "1" or "true"  with the expected(0,1).

**Reply by vadz** [reply]
Oh, thanks, again, I've completely skipped over this part but I see that I need it now and with it I can indeed make it work in exactly the way I want by using
```cpp
    app.add_option("--dsn", [&opts](const CLI::results_t& results)
      {
        if ( results.empty() || results[0].empty() || results[0][0] == '-' )
          throw CLI::RequiredError("database connection string");

        opts.dsn = results[0];
        return true;
      },
      "The database connection string to use"
    )->expected(0, 1);
```
which also means that `--dsn --some-other-option` now gives an error instead of trying to use the other option name as the DSN.

I'm going to wrap this in a function and use it for most (all?) my options taking a value, as I think it provides better user experience than the default behaviour. I wonder if it could be useful to have something like this in the library itself?

In any case, this is definitely good enough for my needs, thanks again for your help!
